File: topics/dynamic-programming/medium/0198-house-robber/solution.cpp
// 0198 · House Robber
// Approach: DP (two variables) | Time: O(n) | Space: O(1)

#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int rob(vector<int>& nums) {
        int prev2 = 0, prev1 = 0;
        for (int num : nums) {
            int curr = max(prev1, prev2 + num);
            prev2 = prev1;
            prev1 = curr;
        }
        return prev1;
    }
};

File: topics/dynamic-programming/medium/0198-house-robber/solution.py
# 0198 · House Robber
# Approach: DP (two variables) | Time: O(n) | Space: O(1)

from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        prev2, prev1 = 0, 0
        for num in nums:
            prev2, prev1 = prev1, max(prev1, prev2 + num)
        return prev1
